File: core/src/main/scala/com/myGames/PokemonInfoScala/AllyPokemon.scala
package com.myGames.PokemonInfoScala

import com.myGames.PokemonInfoScala.AllyPokemon.maxExp


class AllyPokemon(
  override val name: String,
  override val pokemonType: Type.Value,
  override val moveSet: List[Option[Move]],
  var level: Int = 1,
  var exp: Int = 0,
) extends PokemonTrait {

  def gainExp(gainedExp: Int): Unit = {
    exp += gainedExp
    if (exp >= maxExp) levelUp()
  }

  private def levelUp(): Unit = {
    level += 1
    exp = 0
  }

}

object AllyPokemon {
  private val maxExp: Int = 100
}

File: core/src/com/myGames/core/Pokemon/AllyPokemon.java
package com.myGames.core.Pokemon;


public class AllyPokemon extends Pokemon {
  private int maxExp;
  private int level;
  private float exp;

  public AllyPokemon(String givenName, Type givenType, Move[] givenMoveSet) {
    super(givenName, givenType, givenMoveSet);
    maxExp = 100;
    level = 1;
    exp = 0;
  }

  public AllyPokemon(String givenName, Type givenType, Move[] givenMoveSet, int givenLevel, float givenExp) {
    super(givenName, givenType, givenMoveSet);
    maxExp = 100;
    level = givenLevel;
    exp = givenExp;
  }

  public int getLevel() {
    return level;
  }

  public float getExp() {
    return exp;
  }

  public void gainExp(int gainedExp) {
    exp += gainedExp;
    if (exp >= maxExp) levelUp();
  }

  private void levelUp() {
    level += 1;
    exp = 0;
  }

}